### PR TITLE
Fixed a crash in unroller when exiting blocks use switch instead of branch

### PR DESCRIPTION
**Pull request by adam-yang**
The code to structurize loop exits assumed exiting blocks always use BranchInst and never SwitchInst. This caused a crash. This change makes the structurizer preemptively check whether this assumption is true before attempting.

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll [hunk 1]
```diff
@@ -0,0 +1,121 @@
+; RUN: %opt %s -dxil-loop-unroll -S | FileCheck %s
```
Comment (dmpots) [inline tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll:1]: It looks like we could probably simplify this bitcode quite a bit? Would be nice to have a minimal loop with the switch here to demonstrate the problem. There is lots of dxil-isms and debug info that seem unnecessary to trigger the bug.

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp [hunk 2]
```diff
@@ -137,11 +137,13 @@ class DxilLoopUnroll : public LoopPass {
     GetPassOptionUnsigned(O, "MaxIterationAttempt", &MaxIterationAttempt,
                           false);
     GetPassOptionBool(O, "OnlyWarnOnFail", &OnlyWarnOnFail, false);
+    GetPassOptionBool(O, "StructurizeLoopExits", &StructurizeLoopExits, false);
```
Comment (dmpots) [inline lib/Transforms/Scalar/DxilLoopUnroll.cpp:140]: I don't understand why this change is here. I don't see any code reading this option. Is it needed for the fix?

**Review by dmpots — APPROVED** [review]

File: lib/Transforms/Scalar/DxilLoopUnroll.cpp (hunk 2 above)
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:140]: By default, we don't structurize loop exit, and it's controlled by an -opt-enable, which translates to an argument to DxilLoopUnroll pass. When running dxopt.exe, there was no way to turn on the structurizer, so I added this option here. Plus the option should exist anyway for the argument. Without it, our -Odump is actually wrong.

Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:140]: Oops, I just straight up forgot to add it to the test.

Comment (adam-yang) [reply lib/Transforms/Scalar/DxilLoopUnroll.cpp:140]: Fixed now.

File: tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll (hunk 1 above)
Comment (adam-yang) [reply tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll]: Simplified it!

File: tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll [hunk 3]
```diff
@@ -0,0 +1,61 @@
+; RUN: %opt %s -dxil-loop-unroll,StructurizeLoopExits=1 -S | FileCheck %s
```
Comment (pow2clk) [inline tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll:1]: This passes for me without crashing with an unaltered compiler and without the StructurizeLoopExits parameter. Is that expected? From the name of the file, I got the impression that this was going to crash

Comment (adam-yang) [reply tools/clang/test/HLSLFileCheck/hlsl/control_flow/attributes/unroll/unroll_switch_exit_crash.ll:1]: Yes. The crash happens in the structurize code path, which is enabled in dxc.exe through `-opt-enable structurize-loop-exits-for-urnoll` (on by default), but disabled by default on dxopt since there was no way to enable that path without the new parameter.

**Review by pow2clk — APPROVED** [review]
Thanks for the explanation!
